### PR TITLE
Feature: Introduces realtime support.

### PR DESCRIPTION
**Pull request by matthias-mayr**
I'd put this up for discussion as well.

This follows the suggestions of @JimmyDaSilva as discussed in #36. The new functions are close to the ones in libfranka. I would argue that the code is so generic, that threshold of originality is not met and it can be included here.

It is my first time dealing with a preempt_rt kernel and realtime systems, But I had quite some interruptions when connecting via FRI from some 5 year old laptop with an i7. So I spend a bit of time adding this.
I guess there would be more changes needed to introduce actual realtime support. E.g. make this `commanding_status` publisher realltime compliant. This pull request allows edits.

This patch makes realtime support optional. I guess this would be the way to go for this repository.

### CODE REVIEW

**Comment by JimmyDaSilva** [comment]
@matthias-mayr thanks a lot for starting this.
I am quite busy at the moment, but I'll gladly try to help here when I get some time.

> The new functions are close to the ones in libfranka. I would argue that the code is so generic, that threshold of originality is not met and it can be included here.

I agree this should be quite enough.

> E.g. make this commanding_status publisher realltime compliant

Yes, and this is quite important.

These are things I have tried earlier this year and it wasn't enough, unfortunately.
Now that I wrote my own driver for the IIWA, I might have an idea of what the remaining issue would be.
Let's say you specified an FRI control rate of 200Hz on the Java side, the C++ driver is expected to loop with a higher frequency than the specified one, especially if sending torque commands.
This is because KUKA is actually doing something when the robot is in the [kuka::fri::COMMANDING_WAIT:](https://github.com/epfl-lasa/iiwa_ros/blob/master/iiwa_driver/src/iiwa.cpp#L265) state.
Here is some copy/paste from an FRI KUKA example:

>     // If we want to command torques, we have to command them all the time; even in
>     // waitForCommand(). This has to be done due to consistency checks. In this state it is
>     // only necessary, that some torque values are sent. The LBR does not take the
>     // specific value into account.
> 

So I think there should be two loops. One with the specified/desired frequency read from the joint interfaces, and another one faster sending repeatedly the last command read.

**Comment by costashatz** [comment]
@matthias-mayr thanks for this. This looks quite nice. Can you provide some plots of comparison (with/without realtime kernel)? Or at least some quantification (e.g. maximum rate achieved)? I will merge #48 as soon as possible to allow for this PR to proceed.

**Comment by matthias-mayr** [comment]
> Let's say you specified an FRI control rate of 200Hz on the Java side, the C++ driver is expected to loop with a higher frequency than the specified one, especially if sending torque commands.
> This is because KUKA is actually doing something when the robot is in the [kuka::fri::COMMANDING_WAIT:](https://github.com/epfl-lasa/iiwa_ros/blob/master/iiwa_driver/src/iiwa.cpp#L265) state.
> Here is some copy/paste from an FRI KUKA example:
> 
> > ```
> > // If we want to command torques, we have to command them all the time; even in
> > // waitForCommand(). This has to be done due to consistency checks. In this state it is
> > // only necessary, that some torque values are sent. The LBR does not take the
> > // specific value into account.
> > ```
My understanding is that the driver is sending those commands what in COMMANDING_WAIT: https://github.com/epfl-lasa/iiwa_ros/blob/4e5bbe58f4ac0b1b32a405532d6cbdf071a0076a/iiwa_driver/src/iiwa.cpp#L265-L267
`_idle` is set to false and that means that commands are send:
https://github.com/epfl-lasa/iiwa_ros/blob/4e5bbe58f4ac0b1b32a405532d6cbdf071a0076a/iiwa_driver/src/iiwa.cpp#L290-L294
 
> So I think there should be two loops. One with the specified/desired frequency reading from the joint interfaces, and another one faster sending repeatedly the last command read.

My current understanding is that there should be no loop on the ROS side. The FRI receive documentation says:
image
Assuming this is correct, there is no need to introduce another loop.

One thing to pay attention to when relying on the blocking receive function are the parameters set by`setSendPeriodMlliSec` and `setReceiveMultiplier` on the Java side:
image
In the example Java code `setReceiveMultiplier` is set to one 1 and things might break when this is changed.

When removing the ROS loop (which seems to make sense) one should either determine the loop rate (`elapsed_time`) based on the actual time passed between receiving commands or set it based on some parameter like it is now:
https://github.com/epfl-lasa/iiwa_ros/blob/4e5bbe58f4ac0b1b32a405532d6cbdf071a0076a/iiwa_driver/src/iiwa.cpp#L215

since it is used in various places:
https://github.com/epfl-lasa/iiwa_ros/blob/4e5bbe58f4ac0b1b32a405532d6cbdf071a0076a/iiwa_driver/src/iiwa.cpp#L285
https://github.com/epfl-lasa/iiwa_ros/blob/4e5bbe58f4ac0b1b32a405532d6cbdf071a0076a/iiwa_driver/src/iiwa.cpp#L295-L301

tl:dr: I think this line should be removed:
https://github.com/epfl-lasa/iiwa_ros/blob/4e5bbe58f4ac0b1b32a405532d6cbdf071a0076a/iiwa_driver/src/iiwa.cpp#L233

**Comment by JimmyDaSilva** [comment]
> _idle is set to false and that means that commands are send:

You're right! So this should work as is for torque control.

> My current understanding is that there should be no loop on the ROS side.

There is a loop indeed, a `while` loop. What you mean is that you want to remove its fixed frequency by removing the `sleep` function, and relying on the blocking aspect of KUKA's `receive` method.
I think it's a great idea! That is in fact what I ended up doing with my non-ROS driver and it works great.

This is definitely worth a try!

**Comment by matthias-mayr** [comment]
> There is a loop indeed, a while loop. What you mean is that you want to remove its fixed frequency by removing the sleep function, and relying on the blocking aspect of KUKA's receive method.

Yes, of course. Bad wording by me.

I removed the sleep and that allowed to use 333 Hz and 500 Hz. I tried ~30s each and my student is currently running at 500 Hz, so we would know if there are interruptions with the connection when using it more often.

I also tried 1000 Hz. It did not manage to build up a connection when using our newly implemented Cartesian impedance controller. Most likely because the control code takes too long to evaluate.
It did connect using `effort_controllers/JointTrajectoryController`, but it was shaking heavily. My current assumption is that this is due to the PID gains defined there. I can try `effort_controllers/JointGroupEffortController` next.

**Comment by matthias-mayr** [comment]
I tried it today. As expected it did not shake when commanding torques directly. I could only activate the robot for a couple of seconds though. We currently have a load data missmatch that causes ~2.5Nm external torque on joint 1, so it would move when receiving zero torques from ROS. But the FRI connection seems to be open even when the forward button is not pressed.

I also did the following things:
- I wrote some code to [publish the FRI connection quality](https://github.com/matthias-mayr/iiwa_ros/commit/f9a859c02657672ae8e5ca0e0f9f9e13bfc53e67)
- When having the `sleep` still in the loop the connection quality switched between `good` and `excellent`. Both states would allow to command the robot
- Without the sleep it stayed at `excellent`
- I checked at which rate the new `FRIState` messages are published from the control loop and it is indeed 1000 Hz when only relying on the blocking functionality of FRI's `receive`

Given these results, I will add a commit that removes the `sleep` function in the control loop to this PR.

**Comment by costashatz** [comment]
@matthias-mayr what is the status of this?

**Comment by matthias-mayr** [comment]
It works great for us ever since.

**Comment by matthias-mayr** [comment]
Can we merge this? For us this is necessary to achieve high update rates

File: iiwa_driver/src/iiwa.cpp [hunk 1]
```diff
@@ -36,6 +36,33 @@
 #include <thread>
 
 namespace iiwa_ros {
+    bool hasRealtimeKernel() {
```
Comment (costashatz) [inline iiwa_driver/src/iiwa.cpp:39]: Can you make those two global functions in snake case (i.e. with lower case and underscores?) to follow the rest of the styling of the code? Thanks.. Also I believe that those functions need to be inline for better compiler optimization...

**Review by costashatz — CHANGES_REQUESTED** [review]
@matthias-mayr good job. This seems to be almost ready to merge (sorry for the delay). Check one minor comment.. Thanks!

**Comment by matthias-mayr** [comment]
Done
